File: December2022/Java/MaximumProductOfSplittedBinaryTree.java
// @saorav21994

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    long sum = 0;
    static int mod = 1000000007;
    
    // root node contains the sum of its children's values + itself
    private long findSum(TreeNode root) {
        if (root != null) {
            long left = findSum(root.left);
            long right = findSum(root.right);
            root.val += left + right;
            return root.val;
        }
        return 0L;   // null node
    }
    
    private long findProd(TreeNode root) {
        if (root != null) {
            long left = findProd(root.left);
            long right = findProd(root.right);
            long p = root.val * (sum-(root.val));
            // System.out.println("p = " + p);
            return Math.max(p, Math.max(left,right));
        }
        return 0L;
    }
    
    public int maxProduct(TreeNode root) {
        sum = findSum(root);
        long prod = findProd(root);
        // System.out.println(prod);
        return (int)(prod%mod);
    }
}

File: December2022/C++/maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    
    long long findMaxSum( TreeNode*root, long long &ans, long long &totalSum){
        
        if(root == NULL){
            
            return 0;
        }

        long long currSum = findMaxSum(root->left, ans, totalSum) + findMaxSum(root->right, ans, totalSum) + root->val;
        
        ans = max(ans, currSum*(totalSum-currSum));
        
        return currSum;
    }
public:
    int maxProduct(TreeNode* root) {
        
        long long totalSum = 0;
        long long ans = 0;
        long long mod = 1e9 + 7;
        totalSum  = findMaxSum(root, ans, totalSum);
        
        findMaxSum(root, ans, totalSum);
        
        return ans%mod;
    }
};